internal/query/rolebindings.go

```diff
@@ -0,0 +1,844 @@
+		span.SetStatus(codes.Error, err.Error())
+
+		return nil, err
+	}
+
+	// 2. fetch role-binding details for each grant
+	bindings := make([]types.RoleBinding, 0, len(grantRel))
+	errs := make([]error, 0, len(grantRel))
+
+	for _, rel := range grantRel {
+		rbRes, err := e.NewResourceFromIDString(rel.Subject.Object.ObjectId)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		rb, err := e.GetRoleBinding(ctx, rbRes)
+		if err != nil {
+			if errors.Is(err, ErrRoleBindingNotFound) {
+				// print and record a warning message when there's a grant points
+				// to a role-binding that not longer exists.
+				//
+				// this should not happen in normal circumstances, but it's possible
+				// if some role-binding relationships are deleted directly through
+				// spiceDB
+				err := fmt.Errorf("%w: dangling grant relationship: %s", err, rel.String())
+
+				span.RecordError(err)
+				span.SetStatus(codes.Error, err.Error())
+				e.logger.Warnf(err.Error())
+			}
+
+			errs = append(errs, err)
+
+			continue
+		}
+
+		if optionalRole != nil && rb.Role.ID.String() != optionalRole.ID.String() {
+			continue
+		}
+
+		if len(rb.Subjects) == 0 {
+			continue
+		}
+
+		bindings = append(bindings, rb)
+	}
+
+	for _, err := range errs {
+		if err != nil {
+			span.RecordError(err)
+			span.SetStatus(codes.Error, err.Error())
+
+			return nil, err
+		}
+	}
+
+	return bindings, nil
+}
+
+func (e *engine) UpdateRoleBinding(ctx context.Context, actor, rb types.Resource, subjects []types.RoleBindingSubject) (types.RoleBinding, error) {
+	ctx, span := e.tracer.Start(
+		ctx, "engine.UpdateRoleBindings",
+		trace.WithAttributes(
+			attribute.Stringer("rolebinding_id", rb.ID),
+		),
+	)
+	defer span.End()
+
+	dbCtx, err := e.store.BeginContext(ctx)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return types.RoleBinding{}, err
+	}
+
+	if err := e.store.LockRoleBindingForUpdate(dbCtx, rb.ID); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		return types.RoleBinding{}, err
+	}
+
+	rolebinding, err := e.GetRoleBinding(dbCtx, rb)
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		return types.RoleBinding{}, err
+	}
+
+	// 1. find the subjects to add or remove
+	current := make([]string, len(rolebinding.Subjects))
+	incoming := make([]string, len(subjects))
+
+	for i, subj := range rolebinding.Subjects {
+		current[i] = subj.SubjectResource.ID.String()
+	}
+
+	for i, subj := range subjects {
+		incoming[i] = subj.SubjectResource.ID.String()
+	}
+
+	add, remove := diff(current, incoming)
+
+	// return if there are no changes
+	if (len(add) + len(remove)) == 0 {
+		return rolebinding, nil
+	}
+
+	// 2. create relationship updates
+	updates := make([]*pb.RelationshipUpdate, len(add)+len(remove))
+	i := 0
```

Comment:
Why not use `make([]*pb.RelationshipUpdate, 0, len(add)+len(remove))` and then call `append` everywhere? You'll still get only one allocation that way since you're using the `make([]T, length, capacity)` form of `make`.